activitylog/activity.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Abstract Machines
+// SPDX-License-Identifier: Apache-2.0
+
+package activitylog
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/absmach/magistrala/auth"
+	"github.com/absmach/magistrala/internal/apiutil"
+)
+
+type EntityType uint8
+
+const (
+	UserEntity EntityType = iota
+	GroupEntity
+	ThingEntity
+	ChannelEntity
+)
+
+// String representation of the possible entity type values.
+const (
+	userEntityType    = "user"
+	groupEntityType   = "group"
+	thingEntityType   = "thing"
+	channelEntityType = "channel"
+)
+
+var ErrMissingOccurredAt = errors.New("missing occurred_at")
+
+// String converts entity type to string literal.
+func (e EntityType) String() string {
+	switch e {
+	case UserEntity:
+		return userEntityType
+	case GroupEntity:
+		return groupEntityType
+	case ThingEntity:
+		return thingEntityType
+	case ChannelEntity:
+		return channelEntityType
+	default:
+		return ""
+	}
+}
+
+// AuthString returns the entity type as a string for authorization.
+func (e EntityType) AuthString() string {
+	switch e {
+	case UserEntity:
+		return auth.UserType
+	case GroupEntity, ChannelEntity:
+		return auth.GroupType
+	case ThingEntity:
+		return auth.ThingType
+	default:
+		return ""
+	}
+}
+
+// ToEntityType converts string value to a valid entity type.
+func ToEntityType(entityType string) (EntityType, error) {
+	switch entityType {
+	case userEntityType:
+		return UserEntity, nil
+	case groupEntityType:
+		return GroupEntity, nil
+	case thingEntityType:
+		return ThingEntity, nil
+	case channelEntityType:
+		return ChannelEntity, nil
+	default:
+		return EntityType(0), apiutil.ErrInvalidEntityType
+	}
+}
+
+// Query returns the SQL condition for the entity type.
+func (e EntityType) Query() string {
+	switch e {
+	case UserEntity:
+		return "((operation LIKE 'user.%' AND attributes->>'id' = :entity_id) OR (attributes->>'user_id' = :entity_id))"
+	case GroupEntity:
+		return "((operation LIKE 'group.%' AND attributes->>'id' = :entity_id) OR (attributes->>'group_id' = :entity_id))"
+	case ThingEntity:
+		return "((operation LIKE 'thing.%' AND attributes->>'id' = :entity_id) OR (attributes->>'thing_id' = :entity_id))"
+	case ChannelEntity:
+		return "((operation LIKE 'channel.%' AND attributes->>'id' = :entity_id) OR (attributes->>'group_id' = :entity_id))"
```

Comment:
@rodneyosodo 
I could not list `channels`create  event. 
I think this should be `group` or we should change in all events to `channel`